crates/orderbook/src/main.rs

```diff
@@ -538,10 +542,23 @@ async fn main() {
 
         (
             estimator.name(),
-            Box::new(instrumented(instance, estimator.name())),
+            Arc::new(BufferingPriceEstimator::new(Box::new(instrumented(
+                instance,
+                estimator.name(),
+            )))),
         )
     };
 
+    let base_estimators_instances: Rc<RefCell<HashMap<_, _>>> = Default::default();
+    let get_or_create_base_estimator = |estimator: PriceEstimatorType| {
+        base_estimators_instances
+            .clone()
+            .borrow_mut()
+            .entry(estimator)
+            .or_insert_with(|| create_base_estimator(estimator))
+            .clone()
+    };
```

Comment:
Should be able to do
```rust
    let mut base_estimators_instances: HashMap<_, _> = Default::default();
    let mut get_or_create_base_estimator = move |estimator: PriceEstimatorType| {
        base_estimators_instances
            .entry(estimator)
            .or_insert_with(|| create_base_estimator(estimator))
            .clone()
    };
```
to get rid of Rc RefCell. Although for some reason I also had to replace 

.map(get_or_create_base_estimator)

with

 .map(|a| get_or_create_base_estimator(a))

below

Reply:
That's really cool. I was sure there has to be a way to get it working without `Rc` and `RefCell` but I didn't find it.
I think I didn't try `.map(|a| get_or_create_base_estimator(a))`.
I'm not really sure why this makes such a difference. Will do some research on the topic.

Reply:
I was surprised by having to manually use the closure there too. Haven't seen that before. There is a chance it would with 

.map(&mut get_or_create_base_estimator)

but that's still weird.

Reply:
I found out what's going on.
The docs hold the [answer](https://doc.rust-lang.org/std/marker/trait.Copy.html#additional-implementors):
> Closure types, if they capture no value from the environment or if all such captured values implement `Copy` themselves. Note that variables captured by shared reference always implement `Copy` (even if the referent doesn’t), while variables captured by mutable reference never implement `Copy`.

The second sentence is important here. `create_base_estimator()` only captures shared references therefore the closure also implements `Copy`.
`get_or_create_base_estimator()` modifies `base_estimator_instances` which therefore has to be captured by an exclusive reference preventing the closure from implementing `Copy`.
If we take a look at 
```rust
    fn map<B, F>(self, f: F) -> Map<Self, F>
    where
        Self: Sized,
        F: FnMut(Self::Item) -> B,
    {
    // ...
```
we see that it takes an `FnMut` by value and consumes it. If we pass `get_base_estimator` into `.map()` it automatically gets copied and `.map()` only consumes the copy of the closure whereas `get_or_create_base_estimator` would actually be consumed which is why we can only pass it to `.map()` once.